sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md

```diff
@@ -0,0 +1,619 @@
+
+mockClient
+    .Setup(client => client.AcceptNextSessionAsync(
+        It.IsAny<string>(),
+        It.IsAny<ServiceBusSessionReceiverOptions>(),
+        It.IsAny<CancellationToken>()))
+    .ReturnsAsync(mockSessionReceiver.Object);
+
+mockSender
+    .Setup(sender => sender.SendMessageAsync(
+        It.IsAny<ServiceBusMessage>(),
+        It.IsAny<CancellationToken>()))
+    .Returns(Task.CompletedTask);
+
+ServiceBusClient client = mockClient.Object;
+
+List<ServiceBusMessage> testSessionMessages = new();
+Queue<ServiceBusReceivedMessage> messagesToReturn = new();
+
+var mockSessionId = "mySession";
+var numMessagesInSession = 5;
+
+for (var i=0; i<numMessagesInSession; i++)
+{
+    var messageBody = $"message{i}";
+    var messageId = $"messageId{i}";
+    ServiceBusMessage sentMessage = new(messageBody)
+    {
+        SessionId = mockSessionId,
+        MessageId = messageId
+    };
+    testSessionMessages.Add(sentMessage);
+
+    ServiceBusReceivedMessage messageToReceive = ServiceBusModelFactory.ServiceBusReceivedMessage(
+        body: new BinaryData(messageBody),
+        messageId: "messageId2",
+        sessionId: mockSessionId
+        //...
+        );
+    messagesToReturn.Enqueue(messageToReceive);
+}
+
+mockSessionReceiver
+    .Setup(receiver => receiver.ReceiveMessageAsync(
+        It.IsAny<TimeSpan>(),
+        It.IsAny<CancellationToken>()))
+    .ReturnsAsync(messagesToReturn.Dequeue());
+
+BinaryData mockSessionState = new("notSet");
+
+mockSessionReceiver
+    .Setup(receiver => receiver.SetSessionStateAsync(
+        It.IsAny<BinaryData>(),
+        It.IsAny<CancellationToken>()))
+    .Callback<BinaryData, CancellationToken>((st, ct) => mockSessionState = st)
+    .Returns(Task.CompletedTask);
+
+mockSessionReceiver
+    .Setup(receiver => receiver.GetSessionStateAsync(
+        It.IsAny<CancellationToken>()))
+    .ReturnsAsync(() => mockSessionState);
+
+// The rest of this snippet illustrates how to send and receive session messages using the mocked
+// service bus client above, this would be where application methods using sessions would be
+// called.
+
+var mockQueueName = "MockQueueName";
+ServiceBusSender sender = client.CreateSender(mockQueueName);
+
+foreach (ServiceBusMessage message in testSessionMessages)
+{
+    await sender.SendMessageAsync(message);
+}
+
+TimeSpan maxWait = TimeSpan.FromSeconds(30);
+
+ServiceBusSessionReceiver sessionReceiver = await client.AcceptNextSessionAsync(mockQueueName, new ServiceBusSessionReceiverOptions(), CancellationToken.None);
+
+ServiceBusReceivedMessage receivedMessage = await sessionReceiver.ReceiveMessageAsync(maxWait, CancellationToken.None);
+
+BinaryData setState = new("MockState");
+await sessionReceiver.SetSessionStateAsync(setState, CancellationToken.None);
+BinaryData state = await sessionReceiver.GetSessionStateAsync(CancellationToken.None);
+
+Assert.AreEqual(setState, state);
+```
+
+## Sending and Receiving messages using topics and subscriptions
```

Comment:
It may be nice here to also have comments calling out the "this is publishing" and "this is receiving" sections, since the snippet is chonk.

Reply:
nit: we don't use "publishing" anywhere in the SB docs that I know of. Instead, we use "sending"